common/cache/lru.go

```diff
@@ -380,19 +383,39 @@ func (c *lru) putInternal(key interface{}, value interface{}, allowUpdate bool)
 		entry.createTime = c.timeSource.Now()
 	}
 
-	c.byKey[key] = c.byAccess.PushFront(entry)
-	c.updateSizeOnAdd(key, valueSize)
-	for c.isCacheFull() {
-		oldest := c.byAccess.Back().Value.(*entryImpl)
-
-		if oldest.refCount > 0 {
-			// Cache is full with pinned elements
-			// revert the insert and return
-			c.deleteInternal(c.byAccess.Front())
-			return nil, ErrCacheFull
+	// ensuring that the cache has at least one spot for the new entry
+	// different logic between count and size approach
+	if c.isSizeBased() {
+		if valueSize > uint64(c.maxSize()) {
+			// value is too big to be cached, we also don't want to evict everyone else
+			return nil, ErrEntryTooBig
+		}
+		c.byKey[key] = c.byAccess.PushFront(entry)
+		c.updateSizeOnAdd(key, valueSize)
+		for c.isCacheFull() {
+			oldest := c.byAccess.Back().Value.(*entryImpl)
+			if oldest.refCount > 0 {
+				// Cache is full with pinned elements
+				// revert the insert and return
+				c.deleteInternal(c.byAccess.Front())
+				return nil, ErrCacheFull
```

Comment:
Oldest item might be pinned but there may be other items in the cache that are not pinned. Why do we stop when we see a pinned item?